app/controllers/application_controller.rb

```diff
@@ -57,11 +58,45 @@ def no_matching_row e
       error: {
         code: "not_found",
         target: target,
-        message: e.message,
+        message: e.message
       }.compact
     }, status: :not_found
   end
 
+  def foreign_key_constraint_violation e
+    logger.debug "#{e}\n#{e.backtrace.join "\n"}"
+
+    # check if this is a violation of role_memberships_member_id_fkey
+    # or role_memberships_role_id_fkey
+    # sample exceptions:
+    # PG::ForeignKeyViolation: ERROR:  insert or update on table "role_memberships" violates foreign key constraint "role_memberships_member_id_fkey"
+    # DETAIL:  Key (member_id)=(cucumber:group:security-admin) is not present in table "roles".
+    # or
+    # PG::ForeignKeyViolation: ERROR:  insert or update on table "role_memberships" violates foreign key constraint "role_memberships_role_id_fkey"
+    # DETAIL:  Key (role_id)=(cucumber:group:developers) is not present in table "roles".
+    if e.message.index(/role_memberships_member_id_fkey/) ||
+      e.message.index(/role_memberships_role_id_fkey/)
+
+      key_string = ''
+      e.message.split(" ").map do |text|
+        if text["(member_id)"] || text["(role_id)"]
+          key_string = text 
+          break 
+        end 
+      end
+
+      # the member ID is inside the second set of parentheses of the key_string
+      key_index = key_string.index(/\(/, 1) + 1
+      key = key_string[ key_index, key_string.length - key_index - 1 ]
+
+      exc = Exceptions::RecordNotFound.new key, message: "Role #{key} does not exist"
+      record_not_found exc
```

Comment:
The reason why `record_not_found` doesn't see a backtrace is that you are calling `record_not_found` with the exception, but the exception is never raised.

Reply:
One way to handle this would be to have a method `render_record_not_found` that is called from here and from `record_not_found`. That removes the `if` test on the log statement in `record_not_found`, which is good because a future developer isn't going to understand why it's there.